app/interactions/create_match_for_contribution.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+class CreateMatchForContribution < ActiveInteraction::Base
+  object :contribution, class: Listing
+  object :match_with, class: User
+
+  def execute
+    create_match_for_contribution!
+    contribution.matched!
+  end
+
+  private
+
+  def create_match_for_contribution!
+    match_params = if contribution.ask?
+                      { receiver: contribution, provider: create_offer_for_ask! }
+                    elsif contribution.offer? # TODO: check if community resource type when it's added
+                      { receiver: create_ask_for_offer!, provider: contribution }
+                    end
+    Match.create!(match_params.merge(status: 'match_confirmed'))
+  end
+
+  def create_offer_for_ask!
+    Offer.create!(counter_contribution_params)
+  end
+
+  def create_ask_for_offer!
+    Ask.create!(counter_contribution_params)
+  end
```

Comment:
I think this can be simplified somewhat:
```suggestion
  def execute
    Match.create! match_receiver_and_provider.merge(status: 'match_confirmed')
    contribution.matched!
  end

  private

  def match_receiver_and_provider
    if contribution.ask?
      { receiver: contribution, provider: Offer.create!(counter_contribution_params) }
    elsif contribution.offer?
      { receiver: Ask.create!(counter_contribution_params), provider: contribution }
    else
      # TODO: check if community resource type when it's added
      fail "Unhandled contribution type: #{contribution.type}"
    end
  end
```
Mostly just some inlining of methods that, for my eyes, makes the code a tad more readable.

Also note the `else` clause that fails fast instead of silently.

Oh, and i realize you didn't author this code and just moved it into the new interactor -- feel free to ignore this suggestion :)

Reply:
@exbinary I split the logic into multiple methods as the anchor method was getting bloated. But I think I broke it too much.

And I like the `else` clause to fail fast if the contribution type is unknown.